### PR TITLE
ltx: harden weight caches and add serving quality tiers

### PR DESCRIPTION
**Pull request by smartonTT**
## Summary

First slice extracted from `ltx-rt` onto current `main`: LTX weight-cache hardening and serving quality-tier plumbing. Base of the stacked PR1→PR2→PR3 series.

- Content-keyed, source-bound weight caches with publish/reclaim correctness and host weight reuse
- Load watchdog + wall-time anomaly ledger
- bf8 quant configuration and quant-aware cache keys
- high / medium / fast quality-tier environment helpers (`utils/ltx.py`)

Exactly 12 files (per the approved split's PR1 file set); the three SDPA-LoFi presets that self-document as failing the PCC bar are excluded. Extracted from the proven `ltx-rt`→`main` reconciliation branch, so the upstream-`main` reconciliation (mesh-device cache-dir signature, walltime ledger) is carried in.

## Test plan

Host / cache tests — all PASS (2 on a Blackhole Galaxy 4x8 via the device broker):

- `models/tt_dit/tests/unit/test_cache_key.py` — 10 passed, 3 skipped (host-only)
- `models/tt_dit/tests/encoders/gemma/test_gemma_full.py` `[4x8]` — PASSED (device job 206)
- `models/tt_dit/tests/models/ltx/test_connector_cache.py` `[4x8]` — PASSED (device job 208)
- `models/tt_dit/tests/models/ltx/test_prompt_encode_ltx.py` `[4x8]` — PASSED (device job 208)

Girl-singing e2e quality test: the e2e entrypoint (`test_pipeline_ltx_distilled.py`, the `bh_4x8sp1tp0_ring` id) and its `LTX_QUALITY`→bf8-traced wiring (`tests/models/ltx/conftest.py`, `apply_quality_env` callers) are introduced in **PR2**, so the e2e is gated on the PR1+PR2 stack. It is already proven on the full reconciled integration branch (VBench 0.9642 / 0.9660 / 0.9900 / 1.0 / 0.6991, CLIP 30.257) and is re-run on PR2.

### CI Status
_Auto-generated on every push. Badges update live. Click a badge to filter runs by this branch._

- [](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:smarton/ltx-rt-pr1-cache-quality-20260904)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/runtime-sanity-tests.yaml?query=branch:smarton/ltx-rt-pr1-cache-quality-20260904)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:smarton/ltx-rt-pr1-cache-quality-20260904)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/all-model-tests.yaml?query=branch:smarton/ltx-rt-pr1-cache-quality-20260904)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:smarton/ltx-rt-pr1-cache-quality-20260904)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:smarton/ltx-rt-pr1-cache-quality-20260904)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:smarton/ltx-rt-pr1-cache-quality-20260904)

### CODE REVIEW

File: models/tt_dit/utils/cache.py [hunk 1]
```diff
@@ -82,16 +161,18 @@ def load_model(
...
-        model_name=model_name,
-        subfolder=subfolder,
-        parallel_config=parallel_config,
-        mesh_shape=mesh_shape,
-        mesh_device=mesh_device,
-        dtype=dtype,
-        is_fsdp=is_fsdp,
-        required=get_torch_state_dict is None,
-    )
+    key = content_key(tt_model, sources) if sources else None
```
[the remark this answers is not in the document]
Comment (smartonTT) [reply models/tt_dit/utils/cache.py:166]: Verified on this branch. The only transformer `load_model` caller that omits `sources=` is `LTXTransformerCheckpoint.load` at `models/tt_dit/models/transformers/ltx/transformer_ltx.py:1225` — a file in the stacked **PR2** (`models/transformers/ltx/*`), outside this slice. PR1 already threads `sources=` through every caller it owns: the Gemma/connector loads at `models/tt_dit/encoders/gemma/encoder_pair.py:241,291,326`. Content-binding the 22B transformer cache is exactly the `sources=` wiring that lands on that PR2 file; adding it here would edit a PR2 file, which this slice must not do. Resolving as out-of-scope-for-this-slice — the transformer content-bind is carried by PR2.

File: models/tt_dit/utils/cache.py [hunk 2]
```diff
@@ -41,6 +53,67 @@ def cache_dir_is_set() -> bool:
...
+    A HuggingFace hub blob is stored under its own git-lfs sha256, so for the common case the
+    resolved filename *is* an exact content hash and costs one `readlink` — no read of the 46GB
+    file. Anything else falls back to size+mtime, which catches a re-download or a swapped file
+    but not an in-place rewrite that preserves both.
+    """
+    p = Path(path).resolve()
+    st = p.stat()
+    if p.parent.name == "blobs" and len(p.name) == 64 and all(c in "0123456789abcdef" for c in p.name):
+        return f"sha256:{p.name}"
+    return f"stat:{st.st_size}:{st.st_mtime_ns}"
```
[the remark this answers is not in the document]
Comment (smartonTT) [reply models/tt_dit/utils/cache.py]: Fixed in c67746d82c6. The fallback source key now folds `st_ctime_ns` and `st_ino` in beside size+mtime. ctime is kernel-set on any write and cannot be reset from userspace, so a metadata-preserving copy that restores size+mtime still gets a distinct key (its ctime and inode differ). Only source-bound keys change, so no legacy content-blind cache is invalidated.

File: models/tt_dit/utils/cache.py [hunk 3]
```diff
@@ -182,12 +325,63 @@ def _cache_ownership_suffix(mesh_device: ttnn.MeshDevice) -> str:
...
+        if key is not None:
+            parameters = _structure(tt_model)
+            for name, spec in parameters.items():
+                spec["bytes"] = (staging / f"{name}.tensorbin").stat().st_size
+            manifest |= {"content_key": key, "parameters": parameters}
+        (staging / CACHE_DICT_FILE).write_text(json.dumps(manifest, indent=2, sort_keys=True))
+
+        if cache_dir.exists():
+            shutil.rmtree(cache_dir)
+        os.replace(staging, cache_dir)
```
[the remark this answers is not in the document]
Comment (smartonTT) [reply models/tt_dit/utils/cache.py:384]: Fixed in 67c3aef6cdb. `_publish_cache` now re-checks `_cache_is_complete(cache_dir, tt_model, key)` immediately before the destructive rmtree/replace; on a hit it adopts the concurrently-published directory and returns (staging is cleaned by the `finally`). For a content key the winner’s bytes are ours by construction, so a racing publisher no longer deletes a valid cache out from under a concurrent reader.

File: models/tt_dit/utils/cache.py [hunk 4]
```diff
@@ -41,6 +53,67 @@ def cache_dir_is_set() -> bool:
...
+    for name, child in module.named_children():
+        yield from _walk_parameters(child, f"{prefix}{name}.")
+    for name, parameter in module.named_parameters():
+        yield f"{prefix}{name}", parameter
+
+
+def _kernel_capture_only() -> bool:
+    """True during a kernel-prewarm capture pass, which records compile recipes with dispatch off.
+
+    Matches the C++ flag's semantics (kernel_prewarm.cpp): any non-empty value enables it.
```
[the remark this answers is not in the document]
Comment (smartonTT) [reply models/tt_dit/utils/cache.py]: Fixed in 2a5dec3a29a. Verified `grep -rn kernel_prewarm tt_metal/ ttnn/` has zero hits on this branch (the C++ counterpart lands in a later PR of the stack). Reworded the comment to state only the env-var contract the gate actually reads — `TT_METAL_KERNEL_CAPTURE_ONLY`, any non-empty value — dropping the reference to a file absent here.

File: models/tt_dit/tests/models/ltx/test_connector_cache.py [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+from ....encoders.gemma.encoder_pair import _read_connector_checkpoint
+from ....pipelines.ltx.pipeline_ltx import LTXPipeline
+from ....utils import cache
+
+POISONED = os.path.expanduser("~/.cache/tt-dit-ltxrt/ltx-2.3-22b-distilled-1.1")
+
+
+def _ckpt() -> str | None:
+    path = os.environ.get("LTX_CHECKPOINT")
+    return path if path and os.path.exists(path) else None
```
[the remark this answers is not in the document]
Comment (smartonTT) [reply models/tt_dit/tests/models/ltx/test_connector_cache.py]: Fixed in ee0de99de1e. `_ckpt()` now resolves via `default_ltx_checkpoint(...)` and, when that returns a `repo:file` hub ref, falls back to `hf_hub_download(local_files_only=True)` — mirroring `test_pipeline_ltx_distilled._ltx_checkpoint_cached`. The test now runs under HF_HOME CI where weights are hub-staged, not only when `LTX_CHECKPOINT` is set.

File: models/tt_dit/tests/models/ltx/test_prompt_encode_ltx.py [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+)
+
+
+def _maxdiff(a: torch.Tensor, b: torch.Tensor) -> float:
+    return (a.float() - b.float()).abs().max().item()
+
+
+@pytest.mark.skipif(
+    not os.path.exists(default_ltx_checkpoint("ltx-2.3-22b-distilled-1.1.safetensors")),
+    reason="needs the LTX checkpoint (set LTX_CHECKPOINT to a local .safetensors)",
```
[the remark this answers is not in the document]
Comment (smartonTT) [reply models/tt_dit/tests/models/ltx/test_prompt_encode_ltx.py:72]: Fixed in 4d3270a712d. Replaced the `os.path.exists(default_ltx_checkpoint(...))` guard (always false for a hub `repo:file` ref) with a hub-cache-aware `_ltx_checkpoint_cached` that also probes `hf_hub_download(local_files_only=True)`, mirroring the documented distilled-test check. The prompt-independence guard can now gate CI.

File: models/tt_dit/utils/ltx.py [hunk 7]
```diff
@@ -90,6 +90,71 @@ def default_ltx_gemma() -> str:
...
+
+# tier -> (quant, s1_sigmas, s2_sigmas); None keeps the pipeline's shipped default for that knob.
+_QUALITY_TIERS = {
+    "high": (None, None, None),
+    "medium": (FAST_QUANT, FAST_S1_SIGMAS, FAST_S2_SIGMAS),
+    "fast": (FAST_QUANT, FAST_S1_SIGMAS_N3, FAST_S2_SIGMAS),
+}
+
+
+def apply_quality_env(env=None):
```
[the remark this answers is not in the document]
Comment (smartonTT) [reply models/tt_dit/utils/ltx.py:136]: Fixed in ac1b4e842f3. Added `models/tt_dit/tests/unit/test_ltx_quality.py` — 8 host cases covering the high/medium/fast tier mapping, case/whitespace normalization, the invalid-tier `ValueError`, the `LTX_FAST` fallback when `LTX_QUALITY` is unset, the unset no-op, and setdefault precedence. All pass host-only.

File: models/tt_dit/utils/walltime.py [hunk 8]
```diff
@@ -143,19 +149,21 @@ def render(title: str, ledger: _Ledger | None = None, wall: float | None = None)
...
+
+
 def _atexit() -> None:
-    # The ledger surfaces once, at process teardown — under pytest (the tests wire no per-item hook) and
-    # the ``python -m ...`` entrypoints alike.
-    if not _enabled():
+    # pytest gets its per-item/session blocks from conftest hooks; only print here for the
+    # `python -m ...` entrypoints that have no such hook. PYTEST_CURRENT_TEST is cleared by
+    # session end (before this atexit runs), so detect pytest by its imported module instead.
+    if not _enabled() or "pytest" in sys.modules:
```
[the remark this answers is not in the document]
Comment (smartonTT) [reply models/tt_dit/utils/walltime.py:161]: Fixed in bca9af818e0. Ran `test_observability.py` on this branch and confirmed 3 real failures from the walltime rewrite (`_ledger`→`_current`/`_session`, `_ENABLED`→env-read `_enabled()`, pytest-suppressed `_atexit`). Retargeted the fixture and assertions to the split-ledger API and asserted the new contract: under pytest the per-item/session blocks come from the conftest hooks (`pytest_runtest_logreport`/`pytest_sessionfinish`), so `_atexit` stays quiet to avoid double-printing and emits only off the `python -m` path. 8/8 pass.
